### ISSUE TITLE
Loss names are not equal to the outputs dictionary keys of a model, see https://github.com/tensorflow/tensorflow/issues/55435

### ISSUE THREAD

**Issue by szleb** [keras-team/keras#16405]
The original issue with some more details was posted in the tensorflow repo: https://github.com/tensorflow/tensorflow/issues/55435.

**System information**

- Have I written custom code: yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10.0.18363
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.8.0
- Python version: 3.8.10
- Bazel version (if compiling from source): N/A
- GPU model and memory:  None
- Exact command to reproduce: N/A

**The problem**

When creating a model with multiple outputs, that are named by a dictionary, the loss names are not corresponding to the dictionary keys but to the layer names. If the multiple outputs come from the same layer, it is not possible to see which loss belongs to which output.

**Current behavior**

The names for the single losses of a multi output model correspond to the output 
layer names instead to the output names provided by the keys in the outputs dictionary.
I am not sure if this is the desired behavior.

**Expected behavior**

I would expect that the names of the losses are the same as the keys in the dictionary 
passed to the outputs argument when creating the model.

**Source code / logs**

    import numpy as np
    from tensorflow.keras.layers import Dense, Input
    from tensorflow.keras.models import Model
    
    first_input = Input(1)
    x = Dense(100)(first_input)
    x = Dense(1)(x)
    
    second_input = Input(1)
    x2 = Dense(100)(second_input)
    x2 = Dense(1)(x2)
    
    model = Model(
        inputs={"first_input": first_input, "second_input": second_input},
        outputs={"first_output": x, "second_output": x2},
    )
    
    first_input_data = np.random.random(100)
    second_input_data = np.random.random(100)
    first_output_data = np.random.random(100)
    second_output_data = np.random.random(100)
    
    model.compile(loss="mse")
    
    model.fit(
        x={"first_input": first_input_data, "second_input": second_input_data},
        y={"first_output": first_output_data, "second_output": second_output_data},
        batch_size=2,
        epochs=2,
    )

The loss names are visible in the following logging, when running the code example:

    Epoch 1/2
    50/50 [==============================] - 0s 673us/step - loss: 0.2440 - dense_1_loss: 0.1308 - dense_3_loss: 0.1131
    Epoch 2/2
    50/50 [==============================] - 0s 592us/step - loss: 0.1934 - dense_1_loss: 0.0973 - dense_3_loss: 0.0961

**Comment by tilakrayal** [comment]
@gadagashwini ,
I was able to reproduce the issue in tf v2.7, v2.8 and nightly.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/0c80c2cd758f9865c9b4b1a93984d8ef/untitled301.ipynb).

**Comment by itmo153277** [comment]
While I agree that using layer names is a little confusing, I think it is the only consistent way of naming. It is hard to get good names from a nested structure.

Consider these cases:

1. `outputs={"first_output": first_output, "second_output": second_output}` should be `first_output`, `second_output`?
2. `outputs={"nested": {"first_output": first_output}, "second_output": second_output}` should be `nested.first_output`, `second_output`?
3. `outputs={"nested": [first_output, second_output]}` should be `nested[0]`, `nested[1]`?
4. `outputs=[first_output, second_output]` should be `[0]`, `[1]`?

Keep in mind. that 4 is probably the most popular way to describe outputs. Changing naming for this case will break code in many projects (especially for `loss` argument in `compile` function). But if you want to have a different naming way for case 4, imo this only makes everything even more confusing.

**Comment by gadagashwini** [comment]
@szleb,
Did you check @itmo153277  comment.

**Comment by szleb** [comment]
There are two ideas in my mind how to solve the problem:
1) Any dictionary key overwrites the previous name, starting with the name of the layer.
2) All names are concatenated starting from the most outer key to the layer name.

I would prefer the first idea (which was intentionally more or less what I expected), so that the scenarios of @itmo153277 would look like the following:
1. `outputs={"first_output": first_output, "second_output": second_output}` would be `first_output`, `second_output`
2. `outputs={"nested": {"first_output": first_output}, "second_output": second_output}` would be `nested`, `second_output`
3. `outputs={"nested": [first_output, second_output]}` would be `nested`, `nested_1`
4. `outputs=[first_output, second_output]` would be `first_output.node.layer.name`, `second_output.node.layer.name`

Both solutions would not change the naming in case 4.

**Comment by gadagashwini** [comment]
@szleb,
Would you like to contribute for PR to fix this issue.

**Comment by szleb** [comment]
> @szleb, Would you like to contribute for PR to fix this issue.

Yes!

**Comment by tilakrayal** [comment]
@szleb,
As mentioned by the developer, If we have two sets of names to refer to the same objects, it will create the ambiguity to which set you intended to use. 
Please take a look at this [comment](https://github.com/keras-team/keras/pull/16472#issuecomment-1113864309) and respond accordingly. Thank you!
